### PR TITLE
Update C template and introduce programming language features

### PR DESCRIPTION
**Pull request by COM8**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context

The C template was using a quite outdated version of the tester and had some flaws related to potential vulnerabilities.
As we now support a quite large range of programming languages, managing all their features got a bit annoying. This is now simplified in the [ProgrammingLanguageFeatures](https://github.com/ls1intum/Artemis/pull/2150/files#diff-1981327ba5eecfbc95db67961cf2de049944e582c9682aff96d5c9be7cc80d6c).

### Description

I've updated the C template to fix those vulnerabilities and fixed those tests.

Also, the [ProgrammingLanguageFeatures](https://github.com/ls1intum/Artemis/pull/2150/files#diff-1981327ba5eecfbc95db67961cf2de049944e582c9682aff96d5c9be7cc80d6c) now contain the configuration regarding the features of a programming language (e.g. if plagiarism check/sequential test runs/etc. are supported).
The information is present in the server which publishes it to the client using a [ProgrammingLanguageFeatureContributor](https://github.com/ls1intum/Artemis/blob/0cc06eafd85dc46d332d9ee8f53d9205f6801bb5/src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingLanguageFeatureContributor.java). The client uses this information to adjust the import/create dialogs (e.g. by only showing a VHDL exercise in Bamboo-setups, as Jenkins does not support VHDL).

### Steps for Testing

1. Create a new C exercises 
2. The solution should build successfully 
3. Play around with the configuration dialog when creating a new programming exercise: The settings should change (e.g. the package name is only present for Java and Kotlin exercises).
4. Verify that all programming exercises are supported in a Bamboo-environment (TS1/TS3) and that a Jenkins-environment (TS2) does only support Java, Python, C and Haskell.

Peek 2020-10-21 21-55

### CODE REVIEW

**Comment by COM8** [comment]
Looks like I do not have permissions to see why the Java tests fail.

**Comment by sleiss** [comment]
@COM8 I just retriggered the build, seems that a flaky test caused this.

**Comment by krusche** [comment]
@COM8 I added you to the Artemis team. Now you can also deploy and test on the test server. Please sync in Slack with the other developers

**Comment by COM8** [comment]
@krusche Yes, thanks!

**Comment by krusche** [comment]
@COM8 what is the status of this PR? Milestone 4.5.3 is today, and the PR does not seem to be ready to review and merge.
I move it to milestone 4.5.4.

**Comment by krusche** [comment]
regarding the question about the Docker image: we used the same Docker image as for Python programming exercises because both were using the same environment. If you'd prefer two different docker images, we can also setup a new one in our Github account based on com8/artemis-gbs:latest

Let me know what you think

**Comment by krusche** [comment]
You can find the Docker image here: https://github.com/ls1intum/artemis-python-docker/blob/master/Dockerfile
If I see this correctly on https://hub.docker.com/layers/com8/artemis-gbs/latest/images/sha256-1dc2158f02b7b2f21366679be8e3c1c3bf2f9fee857995405d5fd178e0d1d832?context=explore, it is basically the same, so in my opinion there is currently no need to switch it

**Comment by COM8** [comment]
I think, I will set up a new one so once both diverge it is easier to maintain both.

**Comment by krusche** [comment]
> I think, I will set up a new one so once both diverge it is easier to maintain both.

should we then also create this in out `ls1intum` GitHub account? We could give you write access, then you could also maintain it here.

**Comment by krusche** [comment]
Moving this PR to the next milestone.

**Comment by COM8** [comment]
> > I think, I will set up a new one so once both diverge it is easier to maintain both.
> 
> should we then also create this in out `ls1intum` GitHub account? We could give you write access, then you could also maintain it here.

Yes, this would be a good idea.
Yesterday we switched our docker image to fedora with a couple other minor changes. Then I could maintain it.

**Comment by krusche** [comment]
@COM8 just created it here: https://github.com/ls1intum/artemis-c-docker

**Comment by COM8** [comment]
Ready to be reviewed.
I also updated the c-docker-file over at: https://github.com/ls1intum/artemis-c-docker/commit/e9c09a1f0f1f07c20a71cf0be853ff78c4b04a09

What is the name of the docker image I updated over at: https://github.com/ls1intum/artemis-c-docker ?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java:178]: now we have some duplicated code. Can we create a new method: getDockerImage(String programmingLanuage), then the duplication would not be necessary

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java:179]: ls1tum/artemis-c-docker:latest

File: src/main/resources/templates/bamboo/c/sequentialRuns/3_build_and_run_structural_tests.sh
Comment (krusche) [inline src/main/resources/templates/bamboo/c/sequentialRuns/3_build_and_run_structural_tests.sh:3]: if we do not support sequential tests any more, we should make sure that instructors cannot choose this option when they choose C as programming language

**Review by krusche — CHANGES_REQUESTED** [review]
See my inline comments. I am mainly concerned about code duplication and the removal of sequential tests. Please make sure the option is deactivated in the user interface when creating, editing and importing a programming exercises in case the user chooses the programming language C. The server should also check this in `ProgrammingExerciseResource` and reject the generation in case it has been chosen.
This should also be part of a test case then. Thanks!

**Comment by krusche** [comment]
@sleiss @kloessst could you please assist @COM8 for the requested changes? A config which features are supported by which programming languages is also useful when it comes to deactivating options in the user interface in case a template does not support:
* Sequential test runs
* Static code analysis
* Jenkins build plans

I think it is fine if we hard code this information in the server and in the client. I think this could be achieved with a class `ProgrammingExerciseFeatures` with some attributes such as 
* programmingLanguage
* sequentialTestRuns
* staticCodeAnalysis
* bambooBuildSupported
* jenkinsBuildSupported

We could then instantiate one object for each programming language. Based on the values in the object, we would change the configuration of the user interface (client) and change the validation in the `ProgrammingExerciseResource` (server) and potentially also in other places where this information is necessary.

Additional support of features needs changes anyway and then a new release would be necessary. I think we should document somewhere (e.g. in a table in Read The Docs) which programming language template currently supports which features.

We could potentially also work on this in a follow up PR, however this should then follow soon! There are now more and more universities that try out Artemis and that expect consistency in the user interface. I would like to avoid issues here!

**Review by kloessst — COMMENTED** [review]
Checked the Typescript and Java code: Looks good. Having the information about programming exercise feature in one place is a nice idea. Still have to test this.

**Comment by tiffca** [comment]
Tested on test server 1/2. 
C Programming exercise: On test server 1 even when I copy the default solution Makefile and rotX.c file all of the test cases do not pass. Build failed on test server 2 (see below).
Reproduce steps: 
1. Log into test server 1 as test user 1. 
2. View the exercise https://artemistest.ase.in.tum.de/#/courses/304/programming-exercises/4295/code-editor/230980
3. Notice the files are the same as the solution files https://bitbucket.ase.in.tum.de/projects/TIFFTIFFTEST2150/repos/tifftifftest2150-solution/browse

Other than that, building seems to work fine. 

Other exercises: 
TS1: worked fine: Java, Python, kotlin, vhdl
- didn't work: Haskell see test user 1 exercise after copying solution:  https://artemistest.ase.in.tum.de/#/courses/304/programming-exercises/4298/code-editor/230990
- didn't work: assembler, https://artemistest.ase.in.tum.de/#/courses/304/programming-exercises/4301/code-editor/230999

TS2: worked fine (java, python, c, haskell)
- didn't work: haskell - generating the programming exercise gave an internal server error 

Failed to load resource: the server responded with a status of 500 ()
Then it gives the error that the build plans already exist so I have to try to create with new name. 
Gives me the same error. 

- didn't work: C - build failed using solution files. see test user 3 https://artemistest2.ase.in.tum.de/#/courses/1439/programming-exercises/1779/code-editor/6805

**Review by tiffca — CHANGES_REQUESTED** [review]
requesting changes per my last comment (please let me know if I should change to approve!)

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingLanguageFeatureService.java
Comment (kloessst) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingLanguageFeatureService.java:23]: Sometimes a feature is available using Bamboo but not Jenkins. I don't think this is properly represented in this table. E.g. sequential test runs are only available using Bamboo if I'm not mistaken. You could create different Maps for different VCS+CI configurations.

**Review by kloessst — CHANGES_REQUESTED** [review]
Just noticed something, but this might not be an issue:

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingLanguageFeatureService.java
Comment (sleiss) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingLanguageFeatureService.java]: Good point! I changed the `ProgrammingLanguageFeatureService` to an interface and added two implementations (one for Bamboo, one for Jenkins).

**Comment by sleiss** [comment]
@tiffca Thanks for testing! I made an error when defining the docker images: The Assembler template also used the VHDL image, thats why it did not work. Now the build compiles for Assembler (it also passes for the Template, but thats an issue with the template, unrelated to this PR).
The Haskell template gets updated in #2254, thus the errors with the template are not relevant. I will look into fixing the TS2-Haskell issue.

**Review by kloessst — APPROVED** [review]
Code looks good to me now!

**Review by krusche — APPROVED** [review]
The changes look good now, the feature service works. Thanks! Well done!

**Review by fde312 — DISMISSED** [review]
Tested and works fine 👍

**Review by sleiss — DISMISSED** [review]
Tested it on TS1 & TS2 (and wrote parts myself :P)

**Review by kloessst — APPROVED** [review]
Re-Approve

**Review by sleiss — APPROVED** [review]
Re-approve.

**Review by krusche — APPROVED** [review]
re-approve after format fix

**Review by fde312 — APPROVED** [review]

**Comment by krusche** [comment]
👍 :D

**Pull request merged**
